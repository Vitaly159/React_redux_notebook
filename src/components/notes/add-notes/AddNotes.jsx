import { useSelector, useDispatch } from "react-redux";
import { setActiveNote, onDeleteNote, setValueBody, setValueTitle } from "../../../features/addNoteSlice";
import "./addNotes.css";

function AddNotes() {
  const dispatch = useDispatch();
  const notes = useSelector((state) => state.notes.notes);
  const activeNote = useSelector((state) => state.notes.activeNote);
  const searchValue = useSelector((state) => state.notes.searchValue);

  const filteredNotes = notes.filter(
    (note) =>
      note.body.toLowerCase().includes(searchValue.toLowerCase()) ||
      note.title.toLowerCase().includes(searchValue.toLowerCase())
  );

  return filteredNotes.map((note, index) => (
    <div
      key={index}
      className={`added-note ${note.id === activeNote && "active"}`}
      onClick={() => {
        dispatch(setActiveNote(note.id));
        dispatch(setValueTitle(note.title));
        dispatch(setValueBody(note.body));
      }}
    >
      <div className="added-note-title">
        <strong>{note.title}</strong>
        <button className="del" onClick={() => dispatch(onDeleteNote(note.id))}>
          Удалить
        </button>
      </div>

      <p className="added-note-preview">
        {note.body && note.body.substr(0, 30) + "..."}
      </p>
      <small className="note-meta">
        Последнее изменение{" "}
        {new Date(note.lastChanged).toLocaleDateString("ru", {
          hour: "2-digit",
          minute: "2-digit",
        })}
      </small>
    </div>
  ));
}

export default AddNotes;
