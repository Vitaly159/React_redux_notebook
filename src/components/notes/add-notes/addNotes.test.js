import { render, fireEvent, rerender } from "@testing-library/react";
import "@testing-library/jest-dom";
import AddNotes from "./AddNotes";
import { configureStore } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import userEvent from "@testing-library/user-event";
import reducer, { onAddNote, addNoteSlice } from "../../../features/addNoteSlice";
import { act } from "react-dom/test-utils";

const defaultStates = configureStore({ reducer }).getState();

const renderWithRedux = (
  component,
  renderType,
  {
    preloadedState = defaultStates,
    store = configureStore({
      reducer: { notes: reducer },
      preloadedState: { notes: preloadedState },
    }),
  } = {}
) => {
  return {
    ...renderType(<Provider store={store}>{component}</Provider>),
    store,
  };
};

const notesWithLongText = [
  {
    id: 1,
    title: "Задача №1",
    body: "very long text very long text! very long text very long text",
    lastChanged: "Date.now()",
  },
  {
    id: 2,
    title: "Задача №2",
    body: "very long text very long text! very long text very long text",
    lastChanged: "Date.now()",
  },
];

describe("addNotes", () => {
  let component;
  let getByText;
  let store;
  let getByRole;

  beforeEach(() => {
    component = renderWithRedux(<AddNotes />, render, {
      preloadedState: { ...defaultStates, notes: notesWithLongText },
    });
    store = component.store;
    // act(() => store.dispatch(onAddNote("")));
    getByText = component.getByText;
    getByRole = component.getByRole;
  });
  test("Тест onAddNote('')", async () => {
    expect(component.container).toMatchSnapshot();
    // expect(getByText("Имя задачи")).toBeInTheDocument();
    expect(store.getState().notes.notes).toHaveLength(2);
  });
  test("Тест клика по записи для редактирования", async () => {
    const note = getByRole("note-1");
    await userEvent.click(note);
    expect(store.getState().notes.activeNote.length).not.toEqual(0);
    expect(store.getState().notes.valueTitle).toEqual("Задача №1");
    expect(store.getState().notes.valueBody).toEqual(
      "very long text very long text! very long text very long text"
    );
  });
  test("Тест обрезки текста до 30 символов", async () => {
    expect(getByRole("start-text-1")).toBeInTheDocument();
    expect(getByRole("start-text-1")).toHaveTextContent("very long text very long text!...");
  });
  test("Тест кнопки удаления", async () => {
    const buttonRemove = getByRole("btn-rmv-2");
    expect(buttonRemove).toBeInTheDocument();
    await userEvent.click(buttonRemove);
    expect(buttonRemove).not.toBeInTheDocument();
  });
  test("Тест фильтрации", async () => {
    renderWithRedux(<AddNotes />, component.rerender, {
      preloadedState: { ...store.getState().notes, searchValue: "№1"  },
    });
    expect(component.container).toMatchSnapshot();
    expect(getByRole("note-1")).toBeInTheDocument()
    expect(component.queryByText("Задача №2")).not.toBeInTheDocument()
  });
});
