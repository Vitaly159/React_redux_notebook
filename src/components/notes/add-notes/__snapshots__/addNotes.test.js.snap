// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addNotes Тест onAddNote('') 1`] = `
<div>
  <div
    class="added-note false"
    role="note-1"
  >
    <div
      class="added-note-title"
    >
      <strong>
        Задача №1
      </strong>
      <button
        class="del"
        role="btn-rmv-1"
      >
        Удалить
      </button>
    </div>
    <p
      class="added-note-preview"
      role="start-text-1"
    >
      very long text very long text!...
    </p>
    <small
      class="note-meta"
    >
      Последнее изменение
       
      Invalid Date
    </small>
  </div>
  <div
    class="added-note false"
    role="note-2"
  >
    <div
      class="added-note-title"
    >
      <strong>
        Задача №2
      </strong>
      <button
        class="del"
        role="btn-rmv-2"
      >
        Удалить
      </button>
    </div>
    <p
      class="added-note-preview"
      role="start-text-2"
    >
      very long text very long text!...
    </p>
    <small
      class="note-meta"
    >
      Последнее изменение
       
      Invalid Date
    </small>
  </div>
</div>
`;

exports[`addNotes Тест фильтрации 1`] = `
<div>
  <div
    class="added-note false"
    role="note-1"
  >
    <div
      class="added-note-title"
    >
      <strong>
        Задача №1
      </strong>
      <button
        class="del"
        role="btn-rmv-1"
      >
        Удалить
      </button>
    </div>
    <p
      class="added-note-preview"
      role="start-text-1"
    >
      very long text very long text!...
    </p>
    <small
      class="note-meta"
    >
      Последнее изменение
       
      Invalid Date
    </small>
  </div>
</div>
`;
